### PR TITLE
TFRecords exporter

### PR DESCRIPTION
**Pull request by JacopoMangiavacchi**
TensorFlow Records implementation based on explicit use of Protocol Buffer.

I tried different other packages but neither worked out in our scenario so I had to manually define the .proto file and use the protobuf compiler to generate js wrappers to TFRecords-protobuf data structures and reimplement the logic of those packages doing some reverse engineering.

Unfortunately TFRecords documentation is mostly focused only on how to use TFRecords in Python and it wasn't clear how to use TFRecords protobuf datastructures directly.  

Anyway the generated TFRecords files now seems fine even if we definitely need to run some integration test opening these files from Python TF Notebook.

### CODE REVIEW

**Comment by JacopoMangiavacchi** [comment]
NB ProtoBuf file compilation is not integrated into the building process and the .proto file is compiled manually with this command: protoc --proto_path=. --js_out=import_style=commonjs,binary:. tensorFlowRecordsProtoBuf.proto

Unfortunately there's no support for TypeScript on protobuf yet and the generated javascript file got several lint errors if imported from TypeScript.  I had to manually add to this file the following instruction at the beginning: /* eslint-disable */

Anyway the .proto file could be considered a static file and there won't be needs to update this file in the feature even for further customization of the TFRecords exporter logic so I don't think we need to integrate the protoc compilation into the building process.

File: package.json [hunk 1]
```diff
@@ -7,7 +7,10 @@
         "@azure/storage-blob": "^10.3.0",
         "axios": "^0.18.0",
         "bootstrap": "^4.1.3",
+        "buffer-crc32": "^0.2.13",
+        "cryptojs": "^2.5.3",
```
[the remark this answers is not in the document]
Comment (JacopoMangiavacchi) [reply package.json]: I need to export the sha hash of the image buffer. I reused this as in v1 but I’m positive I can replace this package

File: src/providers/export/tensorFlowRecords.ts [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+            recordCrc.setUint32(0, bufferMaskedCRC, true);
+            const bufferDataMaskedCRC = new Buffer(recordCrcBuffer);
+
+            const outBuffer = Buffer.concat([bufferLengthAndMaskedCRC,
+                                             bufferData,
+                                             bufferDataMaskedCRC]);
+
+            await this.storageProvider.writeBinary(fileNamePath, outBuffer);
+
+        } catch (error) {
```
[the remark this answers is not in the document]
Comment (JacopoMangiavacchi) [reply src/providers/export/tensorFlowRecords.ts]: Sure thing. It was just for testing it. I made so many changes here and I definitely need to do some cleanup/refactoring. Btw the CRC package I’m using here is not for crc32C and I’m replacing this with a custom method.

File: src/providers/export/tensorFlowRecords.ts [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+        if (image64.length > 10) {
+            const assetProps = await HtmlFileReader.readAssetAttributesWithBuffer(image64);
+            if (assetProps) {
+                imageInfo.width = assetProps.width;
+                imageInfo.height = assetProps.height;
+            } else {
+                console.log("imageInfo not found");
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (JacopoMangiavacchi) [reply src/providers/export/tensorFlowRecords.ts]: I was thinking more to write a specific integration test using python but in a different PR. Let’s talk offline

File: src/providers/export/tensorFlowRecords.json [hunk 4]
```diff
@@ -1 +1,21 @@
-{}
\ No newline at end of file
+{
+    "type": "object",
+    "title": "TensorFlow Records Settings",
+    "properties": {
+        "assetState": {
```
[the remark this answers is not in the document]
Comment (JacopoMangiavacchi) [reply src/providers/export/tensorFlowRecords.json]: Created a Bug (16837) for add localization to ALL exporter configuration files

File: src/providers/export/tensorFlowRecords.test.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+import MockFactory from "../../common/mockFactory";
+import axios from "axios";
+
+jest.mock("../../services/assetService");
+import { AssetService } from "../../services/assetService";
+
+jest.mock("../storage/localFileSystemProxy");
+import { LocalFileSystemProxy } from "../storage/localFileSystemProxy";
+
+function _base64ToArrayBuffer(base64: string) {
```
[the remark this answers is not in the document]
Comment (JacopoMangiavacchi) [reply src/providers/export/tensorFlowRecords.test.ts]: Removed it.  This method is not used anymore.

File: src/providers/export/tensorFlowRecords.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (JacopoMangiavacchi) [reply src/providers/export/tensorFlowRecords.ts]: As in other exporters that export a single file for each asset we previously decided to simply ignore at the moment individual file export error.  I added some comment about a plan to refactor the exporter interface in another PR in order to return much grained information like how many assets were correctly exported and how many failures.

File: src/providers/export/tensorFlowRecords.ts [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+            const bufferDataMaskedCRC = getInt32Buffer(maskCrc(crc32c(bufferData)));
+
+            // Concatenate all TFRecords Header, Data and Footer buffer
+            const outBuffer = Buffer.concat([bufferLength,
+                                             bufferLengthMaskedCRC,
+                                             bufferData,
+                                             bufferDataMaskedCRC]);
+
+            // Write TFRecords
+            await this.storageProvider.writeBinary(fileNamePath, outBuffer);
```
[the remark this answers is not in the document]
Comment (JacopoMangiavacchi) [reply src/providers/export/tensorFlowRecords.ts]: Sure thing, in the coming next PR

File: src/providers/export/tensorFlowRecords.test.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        expect(exportProvider).not.toBeNull();
+        expect(exportProvider).toBeInstanceOf(TFRecordsJsonExportProvider);
+    });
+
+    describe("Export variations", () => {
+        beforeEach(() => {
+            const assetServiceMock = AssetService as jest.Mocked<typeof AssetService>;
+            assetServiceMock.prototype.getAssetMetadata = jest.fn((asset) => {
+                const mockTag: ITagMetadata = {
+                    name: "Tag 1",
```
Comment (elizabethhalper) [inline src/providers/export/tensorFlowRecords.test.ts]: Not for this PR, but it might be a good idea to add these to the mock factory at some point

**Pull request merged**